Arm64Gen: Convert ARM64Reg to enum class

Dentomologist opened the pull request:
Most changes are just adding ARM64Reg:: in front of the constants.  

Add operator functions to reduce type casting spam

Make DecodeReg return int instead of ARM64Reg

Add EncodeRegTo32 with previous behavior of DecodeReg

Despite appearances, this does not change the value of 
ARM64Reg::INVALID_REG.  The previous value 0xFFFFFFFF was being cast to
an int and wrapping around to -1, this just makes that value explicit.

JosJuice commented on Source/Core/Core/PowerPC/JitArm64/JitArm64_LoadStoreFloating.cpp:
```diff
@@ -164,9 +164,9 @@ void JitArm64::lfXX(UGeckoInstruction inst)
+  regs_in_use[static_cast<size_t>(ARM64Reg::W0)] = 0;
   fprs_in_use[0] = 0;  // Q0
-  fprs_in_use[VD - Q0] = 0;
+  fprs_in_use[static_cast<size_t>(VD - ARM64Reg::Q0)] = 0;
```
Probably preferable:

```suggestion
  fprs_in_use[DecodeReg(VD)] = 0;
```

(Same for all similar lines, like `regs_in_use[DecodeReg(ARM64Reg::W0)] = 0;`)

Dentomologist replied:
I can't follow this suggestion as given, because DecodeReg returns an ARM64Reg which (since it's now an enum class) doesn't implicitly convert to a size_t.  In order to use it as an index BitSet::operator[] I'd either have to cast it to a size_t anyway or write a function to perform that cast, or something else along those lines.

Given that DecodeReg masks its argument to [0,31] and most of the values in question are constants already in that range, the only purpose I can see to using DecodeReg in those cases would be for expressing intent.  Would you say it's worth adding in addition to a cast, or should I just stick with the cast?

As for replacing VD - Q0 with DecodeReg(VD), is VD guaranteed to be in the range [192, 223]?  Its value is set by fpr.RW(), whose behavior is rather opaque to me.

JosJuice replied:
Ah, somehow I thought it was returning an `int` :( Feel free to ignore my suggestion then (either by only using `DecodeReg` on lines that were doing this kind of subtraction in master, or by not using `DecodeReg` at all).

> As for replacing VD - Q0 with DecodeReg(VD), is VD guaranteed to be in the range [192, 223]? Its value is set by fpr.RW(), whose behavior is rather opaque to me.

Yes, in practice it always is. But in case it were not in that range (as you say, it isn't obvious that it is) then using `DecodeReg` would be necessary to get the correct behavior.

JosJuice commented on Source/Core/Core/PowerPC/JitArm64/JitArm64_Integer.cpp:
```diff
@@ -894,8 +894,8 @@ void JitArm64::subfex(UGeckoInstruction inst)
     {
-      MOVI2R(WA, ~i + j, gpr.R(d));
-      ADC(gpr.R(d), WA, WZR);
+      MOVI2R(WA, ~i + j, static_cast<bool>(gpr.R(d)));
```
Bug. The third parameter should just be removed

Dentomologist replied:
Done.

JosJuice commented on Source/Core/Core/PowerPC/JitArm64/JitArm64_Integer.cpp:
```diff
@@ -1403,15 +1403,15 @@ void JitArm64::srawx(UGeckoInstruction inst)
     MOVI2R(WA, 32);
     SUB(WC, WA, WC);
-    LSL(WC, RS, WC);
+    LSL(WC, RS, static_cast<int>(WC));
```
Bug. Should use LSLV (as you will probably find out if you rebase on latest master)

Dentomologist replied:
Done.

Dentomologist commented on Source/Core/Common/Arm64Emitter.h:
```diff
@@ -224,9 +224,65 @@ enum ARM64Reg
   ZR = SP,
 
-  INVALID_REG = 0xFFFFFFFF
+  INVALID_REG = -1,
```
This doesn't actually change the value of INVALID_REG; 0xFFFFFFFF was overflowing the int underlying type and wrapping around to -1.

Dentomologist commented on Source/Core/Common/Arm64Emitter.h:
```diff
@@ -224,9 +224,65 @@ enum ARM64Reg
+constexpr int operator|(const int mask, const ARM64Reg& reg)
+{
+  return static_cast<int>(reg) | mask;
+}
+constexpr int operator|(const ARM64Reg& lhs, const ARM64Reg& rhs)
+{
+  return static_cast<int>(lhs) | static_cast<int>(rhs);
+}
+constexpr ARM64Reg operator+(const ARM64Reg& reg, const int addend)
+{
+  return static_cast<ARM64Reg>(static_cast<int>(reg) + addend);
+}
+constexpr ARM64Reg operator+(const int addend, const ARM64Reg& reg)
+{
+  return static_cast<ARM64Reg>(static_cast<int>(reg) + addend);
+}
+constexpr ARM64Reg operator+(const ARM64Reg& lhs, const ARM64Reg& rhs)
+{
+  return static_cast<ARM64Reg>(static_cast<int>(lhs) + static_cast<int>(rhs));
+}
+constexpr ARM64Reg operator-(const ARM64Reg& reg, const int addend)
+{
+  return static_cast<ARM64Reg>(static_cast<int>(reg) - addend);
+}
+constexpr ARM64Reg operator-(const int addend, const ARM64Reg& reg)
+{
+  return static_cast<ARM64Reg>(addend - static_cast<int>(reg));
+}
+constexpr ARM64Reg operator-(const ARM64Reg& lhs, const ARM64Reg& rhs)
+{
+  return static_cast<ARM64Reg>(static_cast<int>(lhs) - static_cast<int>(rhs));
+}
+constexpr int operator<<(const ARM64Reg& reg, const int distance)
+{
+  return static_cast<int>(reg) << distance;
+}
+constexpr int operator>>(const ARM64Reg& reg, const int distance)
+{
+  return static_cast<int>(reg) >> distance;
+}
```
I don't like the inconsistency of returning an int for the bitwise and shift operators and ARM64Reg for + and -, but the point of these operators is to reduce line noise by getting rid of a bunch of casts, and those return types reflect the expected type in most of those usages.

JosJuice replied:
I think this is reasonable for the most part, but why would someone want to add an (or subtract) an ARM64Reg with an ARM64Reg? ARM64Reg and int makes sense to me, but not two ARM64Reg.

Dentomologist replied:
Adding two ARM64Regs is done in MOVI2FDUP and EncodeRegToSingle.  I don't think anything actually calls MOVI2FDUP so we could remove it, and if we made DecodeReg return an int (see other comment for more on the subject) then EncodeRegToSingle wouldn't need to add two ARM64Regs either.

Subtracting two registers was only done in the context of statements like fprs_in_use[VD - Q0], but since I'm replacing those anyway I suppose I could get rid of the subtraction operators.

Dentomologist commented on Source/Core/Core/PowerPC/JitArm64/JitArm64_RegCache.cpp:
```diff
@@ -744,7 +744,7 @@ BitSet32 Arm64FPRCache::GetCallerSavedUsed() const
   {
     if (it.IsLocked())
-      registers[it.GetReg() - Q0] = true;
+      registers[GetUsedRegisterIndex(it.GetReg())] = true;
```
I believe the values in m_host_registers will all be Q registers which have values [192, 223].  Values in this range are mapped to [0, 31] both before and after this change.

JosJuice replied:
Yes, this change is correct. The function name `GetUsedRegisterIndex` sounds a bit unwieldy, though... Now that I think about it, is there actually any reason why `DecodeReg` shouldn't return an `int`?

Dentomologist replied:
There are a number of places where DecodeReg is called and the return value is passed into another function that expects an ARM64Reg, so I wouldn't want to get rid of it.

On the other hand, there are a number of places where the return value is used in bitwise operations which effectively treat it as an int.

I'm not sure if this is best done in this PR or moved to another one, but we may want to create another function and have one return an Arm64Reg and the other an int.  I don't know what we would name them; DecodeReg already sounds a bit opaque to me even though I know what it does, so that might be an opportunity to give it a clearer name, but perhaps that name already is clear to people like you who are familiar with ARM assembly.

JosJuice replied:
> There are a number of places where DecodeReg is called and the return value is passed into another function that expects an ARM64Reg, so I wouldn't want to get rid of it.

After a quick look (I may have missed some instances), the functions I could find that were being called with the return value of `DecodeReg` were `UnlockRegister` (from JitArm64_RegCache) and `MOV`/`SXTW` (from JitArm64_Integer.cpp). In both of these cases, the `DecodeReg` call doesn't make sense and should just be removed as far as I can tell. (Things end up working anyway because the register in question is always a 32-bit GPR.)

Dentomologist replied:
I started converting DecodeReg to return an int but found the change ended up being much more extensive than anticipated.

JitArm64::JumpIfCRFieldBit: calls DecodeReg and passes the result to either CBZ or CBNZ, which each insert the reg into a FixupBranch which is returned up to the bcx and bclrx functions, which pass the FixupBranch to SetJumpTarget, where the reg value is decoded a couple times and also passed to Is64Bit.

JitArm64::mcrxr: 
	calls LDRB, which calls one of two overloads of EncodeLoadStoreIndexedInst, which each call Is64Bit, IsVector, and DecodeReg. 
	calls ADD, which calls EncodeArithmeticInst, which calls Is64Bit and DecodeReg
	Initializes ArithOption, which initializes a member variable with it.

There are similar chains for crXXX, mfcr, mtcrf, cmpl, and cmpli in JitArm64, among others I haven't taken the time to investigate.

It would be possible to make all those changes, but it starts to approach the point of getting rid of ARM64Reg entirely.

JosJuice replied:
At least for `JumpIfCRFieldBit` and `mcrxr`, the appropriate thing to do would be to create a function `EncodeRegTo32` which works identically to how `DecodeReg` currently works. (Decoding a register and converting a register to its 32-bit form are two rather distinct operations semantically, but because the enum happens to start with `W0`, `reg & 0x1F` can be used to implement both.)

Also, a correction: In my previous comment, I didn't take into account that CR registers normally are mapped to 64-bit GPRs. Instead of removing the usages of `DecodeReg` that I mentioned in that comment, please replace them with `EncodeRegTo32`.

JosJuice replied:
Or in other words: Let's have both `DecodeReg` and `EncodeRegTo32`, with the former returning `int` and the latter returning `ARM64Reg`.

Dentomologist commented on Source/Core/Common/Arm64Emitter.cpp:
```diff
@@ -1223,44 +1177,44 @@ void ARM64XEmitter::_MSR(PStateField field, ARM64Reg Rt)
   ASSERT_MSG(DYNA_REC, Is64Bit(Rt), "MSR: Rt must be 64-bit");
   GetSystemReg(field, o0, op1, CRn, CRm, op2);
-  EncodeSystemInst(o0, op1, CRn, CRm, op2, DecodeReg(Rt));
+  EncodeSystemInst(o0, op1, CRn, CRm, op2, Rt);
```
Rt is decoded inside EncodeSystemInst, so there's no need to do it here.

Dentomologist commented on Source/Core/Common/Arm64Emitter.h:
```diff
@@ -256,17 +268,21 @@ constexpr bool IsGPR(ARM64Reg reg)
 constexpr ARM64Reg EncodeRegToSingle(ARM64Reg reg)
 {
-  return static_cast<ARM64Reg>(DecodeReg(reg) + S0);
+  return static_cast<ARM64Reg>(ARM64Reg::S0 | DecodeReg(reg));
```
S0 = 64 and DecodeReg return value is in range [0, 31] so these are equivalent.

Dentomologist commented:
As per previous discussion:

Made DecodeReg return int, and added EncodeRegTo32 which returns ARM64Reg
Replaced DecodeReg with EncodeRegTo32 where appropriate
Removed GetUsedRegisterIndex function and replaced calls with DecodeReg
Removed extraneous operator functions

Dentomologist commented:
One last thing before I remove draft status: this builds properly and isn't supposed to change any functionality, but since I don't have an ARM device to test it myself it would be helpful for someone who does to run a few games as a sanity check.

JosJuice commented:
I did some quick tests of a few games. Things seem to be working fine.

JosJuice reviewed (approved):
Code changes LGTM, but you should probably remove the lines of the commit message that come after "Most changes are just adding ARM64Reg:: in front of the constants."

Dentomologist commented:
> Code changes LGTM, but you should probably remove the lines of the commit message that come after "Most changes are just adding ARM64Reg:: in front of the constants."

Done, and draft status removed.

JosJuice commented:
Gonna take the opportunity to merge this before someone merges one of my open JitArm64 PRs and causes a merge conflict. degasus was the one who suggested making this into an enum class so I'm going to assume he doesn't mind this being merged
